src/Exercise/PhpPromotion.php

```diff
@@ -0,0 +1,163 @@
+    }
+
+    public function getInitialCode(): SolutionInterface
+    {
+        return SingleFileSolution::fromFile(__DIR__ . '/../../exercises/php-promotion/initial/php-7-class.php');
+    }
+
+    public function getType(): ExerciseType
+    {
+        return ExerciseType::CLI();
+    }
+
+    public function getArgs(): array
+    {
+        return [];
+    }
+
+    public function check(Input $input): ResultInterface
+    {
+        $statements = $this->parser->parse((string) file_get_contents($input->getRequiredArgument('program')));
+
+        /** @var Class_|null $node */
+        $className = (new NodeFinder())->findFirstInstanceOf($statements, Class_::class)?->name->name;
+
+        if (null === $className) {
+            return Failure::fromNameAndReason($this->getName(), 'No class was found');
+        }
+
+        (static fn () => require $input->getRequiredArgument('program'))();
+
+        $reflectionClass = new ReflectionClass($className);
+        $params = collect($reflectionClass->getMethod('__construct')->getParameters());
+        $params = $params->flatMap(fn (\ReflectionParameter $prop) => [$prop->getName() => $prop]);
+
+        if ($missing = array_diff(['visitor', 'key', 'config'], $params->keys()->getArrayCopy())) {
+            return Failure::fromNameAndReason($this->getName(), pluralise(
+                'Property "%s" was missing',
+                $missing,
+                implode('" & "', $missing)
+            ));
+        }
+
+        $promoted = $params->filter(fn (\ReflectionParameter $prop) => $prop->isPromoted());
+
+        if ($notPromoted = array_diff(['visitor', 'key'], $promoted->keys()->getArrayCopy())) {
+            return Failure::fromNameAndReason($this->getName(), pluralise(
+                'Property "%s" was not promoted',
+                $notPromoted,
+                implode('" & "', $notPromoted)
+            ));
+        }
+
+        if ($promoted->get('config')) {
+            return Failure::fromNameAndReason($this->getName(), 'Property "config" should not be promoted');
+        }
+
+        $expectedProperties = [
+            'basePath' => 'protected',
+            'key' => 'private',
+            'visitor' => 'private',
+        ];
+
+        $properties = collect($reflectionClass->getProperties());
+        $properties = $properties->flatMap(fn (\ReflectionProperty $prop) => [
+            $prop->getName() => $this->getPropertyVisibility($prop)
+        ])->getArrayCopy();
+        ksort($properties);
+
+        if ($changedVisibility = array_keys(array_diff_assoc($expectedProperties, $properties))) {
+            return Failure::fromNameAndReason($this->getName(), pluralise(
+                'Visibility changed for property "%s"',
+                $changedVisibility,
+                implode('" & "', $changedVisibility)
+            ));
+        }
+
+        $types = $properties->map(fn (\ReflectionProperty $prop) => $prop->getType()?->getName());
```

Comment:
`$properties` is an array here - maybe you wanna add the `ksort` method to the collection class?

Reply:
Yeah you're right I should